c++/src/kj/exception.c++

```diff
@@ -254,9 +255,138 @@ ArrayPtr<void* const> getStackTrace(ArrayPtr<void*> space, uint ignoreCount) {
+      close(out[1]);
+      return Subprocess({ .pid = pid, .in = in[1], .out = out[0] });
+    } else {
+      // child
+      close(in[1]);
+      close(out[0]);
+
+      // redirect stdin
+      close(0);
+      if(dup(in[0]) < 0) {
+        _exit(2 * errno);
+      }
+
+      // redirect stdout
+      close(1);
+      if(dup(out[1]) < 0) {
+        _exit(3 * errno);
+      }
+
+      execl("/bin/sh", "sh", "-c", cmd.cStr(), nullptr);
+      _exit(1);
+    }
+  }
+
+  int closeAndWait() {
+    close(in);
+    close(out);
+    int status;
+    waitpid(pid, &status, 0);
+    return status;
+  }
+
+  int pid;
+  int in;
+  int out;
+};
+
+
+static kj::Maybe<kj::String> tryLlvmSymbolizer(ArrayPtr<void* const> trace) {
+  const char* llvmSymbolizer = getenv("LLVM_SYMBOLIZER");
+  if (llvmSymbolizer == nullptr) {
+    return nullptr;
+  }
+
+  auto cmd = kj::str(llvmSymbolizer, " --pretty-print --relativenames");
+  KJ_IF_MAYBE(subprocess, Subprocess::exec(cmd)) {
+    // write addresses as "CODE <file_name> <hex_address>" lines.
+    auto addrs = strArray(KJ_MAP(addr, trace) {
+      #if KJ_HAS_LIBDL
+        Dl_info info;
+        if (dladdr(addr, &info)) {
+          uintptr_t offset = reinterpret_cast<uintptr_t>(addr) -
+                              reinterpret_cast<uintptr_t>(info.dli_fbase);
+          return kj::str("CODE ", info.dli_fname, " 0x", reinterpret_cast<void*>(offset));
+        } else {
+          return kj::str("CODE 0x", reinterpret_cast<void*>(addr));
+        }
+      #else
+        return kj::str("CODE 0x", reinterpret_cast<void*>(addr));
+      #endif
+    }, "\n");
+    if (write(subprocess->in, addrs.cStr(), addrs.size()) != addrs.size()) {
+      KJ_LOG(ERROR, "write error", strerror(errno));
+      return nullptr;
+    }
+    close(subprocess->in);
+
+    // read result
+    auto out = fdopen(subprocess->out, "r");
+    if (!out) {
+      KJ_LOG(ERROR, "fdopen error", strerror(errno));
+      return nullptr;
+    }
+
+    kj::String lines[256];
+    size_t i = 0;
+    for (char line[512]; fgets(line, sizeof(line), out) != nullptr;) {
+      if (i < kj::size(lines)) {
+        lines[i++] = kj::str(line);
+      }
+    }
+    int status = subprocess->closeAndWait();
+    if (WIFEXITED(status)) {
+      if (WEXITSTATUS(status) != 0) {
+        KJ_LOG(ERROR, "bad exit code", WEXITSTATUS(status));
+        return nullptr;
+      }
+    } else {
+      KJ_LOG(ERROR, "bad exit status", status);
+      return nullptr;
+    }
+    return kj::strArray(kj::arrayPtr(lines, i), "");
+  } else {
+    return nullptr;
+  }
+}
+#endif
+
 String stringifyStackTrace(ArrayPtr<void* const> trace) {
   if (trace.size() == 0) return nullptr;
-  if (getExceptionCallback().stackTraceMode() != ExceptionCallback::StackTraceMode::FULL) {
+  auto stackTraceMode = getExceptionCallback().stackTraceMode();
+#if (__linux__ || __APPLE__ || __CYGWIN__) && !__ANDROID__
+  if (stackTraceMode == ExceptionCallback::StackTraceMode::TRY_LLVM_SYMBOLIZER) {
```

Comment:
Hmm. If it's necessary to set a custom stack trace mode anyway (requiring code added to programs that use this), maybe we should instead provide a way to specify a callback to symbolize stack traces, rather than add the llvm-symbolizer-specific logic to KJ?

Perhaps we could even use that rust addr2line library in-process rather than shelling out.

It would be especially nice if we could make it so simply linking against that library will cause KJ to use it, without any other code changes needed in the binary. That way we can link it into all our KJ tests...

(Another possible path is to automatically assume `TRY_LLVM_SYMBOLIZER` here but I worry there are security implications to us happily invoking whatever binary happens to be specified by some environment variable...)